# COT_lab — Kaggle pipeline (modular)

Each training condition is a **self-contained block**: train → infer → prune checkpoint. If one block dies, others are unaffected. Every cell is resumable across the 12-hour Kaggle session boundary.

**Layout**
1. Setup (clone, install, optional git auth) — run once per session.
2. Stages 1 & 2 (download + filter) — run once per fresh checkout.
3. **Base sweep** (FLAN-T5-base): direct_ft, set_b, set_c, set_a, baseline, set_c_mix.
4. **Large crossover** (FLAN-T5-large, optional): direct_ft, set_b, set_c, set_a.
5. Stage 5a accuracy refresh (rebuilds the full table).
6. Stage 5b ReCEval (recommended in a separate session — slow).
7. Push results + aggressive cleanup.

**Storage hygiene (Kaggle 20 GB limit)**
After each block's inference completes, a `cleanup_ckpts` cell prunes the run to its best checkpoint only (~50% reduction on T5-base, ~66% on T5-large). After Stages 4 and 5 are fully done, you can drop the best checkpoint too — generations are on disk and pushable.

**Resumability**
- Training: re-run the cell with `--resume`. The trainer picks up from the latest checkpoint and walks the remaining epochs.
- Inference: re-run the cell. Already-written rows are detected by line count and skipped.
- Evaluation: idempotent — re-running rebuilds the CSV.

**Rough runtime per block on a free-tier T4**

| Block | Train | Infer | Prune | Total |
|---|---:|---:|---:|---:|
| direct_ft (base) | 15–30 min | ~10 min | <1 min | ~30–40 min |
| set_b (base) | 25–45 min | ~10 min | <1 min | ~35–55 min |
| set_c (base) | 20–45 min | ~10 min | <1 min | ~30–55 min |
| set_a (base) | 60–120 min | ~10 min | <1 min | ~70–130 min |
| baseline | — | ~10 min | — | ~10 min |
| set_c_mix (base) | 20–45 min | ~10 min | <1 min | ~30–55 min |
| direct_ft (**large**) | ~60–90 min | ~25 min | <1 min | ~90 min |
| set_b (**large**) | ~45–75 min | ~25 min | <1 min | ~75 min |
| set_c (**large**) | ~45–75 min | ~25 min | <1 min | ~75 min |
| set_a (**large**) | ~60–90 min | ~25 min | <1 min | ~90 min |
| Stage 5b ReCEval (all) | — | — | — | ~3–5 h |

## 1. Setup

In [ ]:
# ── EDIT THESE ───────────────────────────────────────────────────────────────
GH_USER = 'rihembenabdallah18'
GH_REPO = 'LAB1'
BRANCH  = 'main'
# ─────────────────────────────────────────────────────────────────────────────

REPO_URL = f'https://github.com/{GH_USER}/{GH_REPO}.git'
REPO_DIR = f'/kaggle/working/{GH_REPO}'

import os, subprocess
if not os.path.isdir(REPO_DIR):
    subprocess.check_call(['git', 'clone', '--branch', BRANCH, REPO_URL, REPO_DIR])
else:
    subprocess.check_call(['git', '-C', REPO_DIR, 'fetch', 'origin'])
    subprocess.check_call(['git', '-C', REPO_DIR, 'checkout', BRANCH])
    subprocess.check_call(['git', '-C', REPO_DIR, 'pull', '--rebase', 'origin', BRANCH])

os.chdir(REPO_DIR)
print('cwd:', os.getcwd())

Already on 'dev'


M	scripts/01_download.sh
M	scripts/02_filter.sh
M	scripts/03_train_direct_ft.sh
M	scripts/03_train_set_a.sh
M	scripts/03_train_set_b.sh
M	scripts/03_train_set_c.sh
M	scripts/04_inference.sh
M	scripts/05a_accuracy.sh
M	scripts/05b_receval.sh
M	scripts/status.sh
Your branch is up to date with 'origin/dev'.
Already up to date.
cwd: /kaggle/working/COT_lab


From https://github.com/rihembenabdallah18/COT_lab
 * branch            dev        -> FETCH_HEAD


In [27]:
!pip install -q -r requirements.txt
# Kaggle pre-installs a newer `peft` that expects EncoderDecoderCache from
# transformers >= 4.43; we pin transformers==4.42.4 for reproducibility. We
# don't use peft anywhere — uninstall it so transformers' Trainer skips the
# optional PEFT import path cleanly.
!pip uninstall -y -q peft
!python -m spacy download en_core_web_sm -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 kB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 4.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 779.1/779.1 MB 2.1 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.3/9.3 MB 4.4 MB/s eta 0:00:000:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 547.8/547.8 kB 541.8 kB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 314.1/314.1 kB 22.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 65.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.5/6.5 MB 120.2 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 101.5 MB/s eta 0:00:0000:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.0/18.0 MB 88.1 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 725.0/725.0 kB 44.5 MB/s eta 0:

In [2]:
import torch
print('cuda :', torch.cuda.is_available())
if torch.cuda.is_available():
    print('gpu  :', torch.cuda.get_device_name(0))
    print('vram :', round(torch.cuda.get_device_properties(0).total_memory / 1e9, 1), 'GB')

cuda : True
gpu  : Tesla T4
vram : 15.6 GB


### 1.1 Configure GitHub push (optional)

Skip if you only want to **read** results locally. Required for the push cells at the end of the notebook.

1. Create a GitHub Personal Access Token with `repo` scope.
2. Kaggle → Settings → Add-ons → Secrets → add `GITHUB_TOKEN`.
3. Run this cell.

In [40]:
import os, subprocess

TOKEN = None
try:
    from kaggle_secrets import UserSecretsClient
    TOKEN = UserSecretsClient().get_secret('GITHUB_TOKEN')
except Exception:
    TOKEN = os.environ.get('GITHUB_TOKEN')

if not TOKEN:
    raise RuntimeError(
        'Set Kaggle secret GITHUB_TOKEN (Settings -> Add-ons -> Secrets) '
        'or export the env var GITHUB_TOKEN before running this cell.'
    )

auth_url = f'https://{GH_USER}:{TOKEN}@github.com/{GH_USER}/{GH_REPO}.git'
subprocess.check_call(['git', '-C', REPO_DIR, 'remote', 'set-url', 'origin', auth_url])
subprocess.check_call(['git', '-C', REPO_DIR, 'config', 'user.email', 'kaggle@cot-lab.local'])
subprocess.check_call(['git', '-C', REPO_DIR, 'config', 'user.name',  'Kaggle Runner'])
print('git auth configured for', GH_USER + '/' + GH_REPO)

git auth configured for rihembenabdallah18/COT_lab


## 2. Stage 1 — Download GSM8K + Ho et al. teacher CoTs

~10-20 min on first run; cached afterwards (skips on re-run).

In [3]:
!bash scripts/01_download.sh

[gsm8k] downloading via datasets.load_dataset('gsm8k', 'main')
README.md: 7.93kB [00:00, 15.8MB/s]
main/train-00000-of-00001.parquet: 100%|███| 2.31M/2.31M [00:00<00:00, 4.23MB/s]
main/test-00000-of-00001.parquet: 100%|██████| 419k/419k [00:00<00:00, 3.20MB/s]
Generating test split: 100%|█████| 1319/1319 [00:00<00:00, 307042.23 examples/s]
[gsm8k] saved 7473 train, 1319 test -> /kaggle/working/COT_lab/data/raw/gsm8k
[ho] downloading shared folder zip (~924 MB) to /tmp/tmplnflvrf8/release.zip
[curl] curl -fSL --retry 3 -o /tmp/tmplnflvrf8/release.zip https://www.dropbox.com/sh/hwcncpyomx87h20/AACqgVdd-ZzBQ3ncJcKqw0cVa?dl=1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   140  100   140    0     0    690      0 --:--:-- --:--:-- --:--:--   689
100    17  100    17    0     0     18      0 --:--:-- --:--:-- --:--:--    18
100  880M  100  880M    0     0  80.4M      0  0:00:10

## 3. Stage 2 — Build training sets (A / B / C / Direct FT)

<1 min. Run whenever the calculator or filter logic changes. Prints set sizes and the number of calculator edits applied to Set C.

In [4]:
!bash scripts/02_filter.sh

GSM8K train rows: 7473
  Direct FT      : 7473 -> /kaggle/working/COT_lab/data/processed/direct_ft.jsonl
  Set A (no flt) : 7473 -> /kaggle/working/COT_lab/data/processed/set_a_nofilter.jsonl
  Set B (Magist) : 3389 -> /kaggle/working/COT_lab/data/processed/set_b_magister.jsonl
  Set C (calc.)  : 3389 -> /kaggle/working/COT_lab/data/processed/set_c_calculator.jsonl
  skipped: no_teacher=0 unparseable_gold=0 unparseable_teacher_pred=78
  Set B keep rate (of A): 45.3%
  Set C keep rate (of A): 45.3%
  CoTs the calculator edited: 187 / 7473

--- 2 example records, Set A ---
{
  "sample_index": 0,
  "question": "Natalia sold clips to 48 of her friends in April, and then she sold half as many clips in May. How many clips did Natali...",
  "cot": "In April, Natalia sold 48 clips. \nIn May, she sold half as many clips. Half of 48 is 24, so she sold 24 clips in May. \n\n...",
  "gold_answer": "72",
  "teacher_predicted_answer": "72."
}
{
  "sample_index": 1,
  "question": "Weng earns $12 an ho

---

## 4. Base model sweep — FLAN-T5-base

Five independent blocks. Run them in any order; each writes to its own `outputs/checkpoints/{cond}/` and `outputs/generations/{cond}.jsonl`. Order suggested below puts cheapest first so you get early signal.

**If a block fails** (OOM, timeout, etc.): just re-run its cells — train resumes from the last checkpoint, inference resumes from the line count.

### 4.student_direct_ft — `student_direct_ft`

Cheapest. Targets collapse to `#### {answer}`. Ho et al.'s reference *standard fine-tune* condition (~5.08% on GSM8K test).

In [ ]:
# Train. Resumable: re-run with --resume after a timeout.
!python -m src.train.finetune \
  --config config/config.yaml \
  --train data/processed/direct_ft.jsonl \
  --run-name student_direct_ft
  # add --resume after a timeout

In [ ]:
# Inference on the GSM8K test set. Resumable.
!python -m src.inference.generate --condition student_direct_ft

In [ ]:
# Prune to the best checkpoint only (keeps disk under control).
!python -m src.utils.cleanup_ckpts --run-name student_direct_ft --keep best

### 4.student_set_b — `student_set_b`

Magister filter: teacher CoTs whose final answer matches gold (~3,389 rows).

In [ ]:
# Train. Resumable: re-run with --resume after a timeout.
!python -m src.train.finetune \
  --config config/config.yaml \
  --train data/processed/set_b_magister.jsonl \
  --run-name student_set_b
  # add --resume after a timeout

In [ ]:
# Inference on the GSM8K test set. Resumable.
!python -m src.inference.generate --condition student_set_b

In [ ]:
# Prune to the best checkpoint only (keeps disk under control).
!python -m src.utils.cleanup_ckpts --run-name student_set_b --keep best

### 4.student_set_c — `student_set_c`

Same membership as Set B, with calculator-patched intermediate arithmetic. Tests whether cleaner math in training data helps the student.

In [8]:
# Train. Resumable: re-run with --resume after a timeout.
!python -m src.train.finetune \
  --config config/config.yaml \
  --train data/processed/set_c_calculator.jsonl \
  --run-name student_set_c
  # add --resume after a timeout

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.
0it [00:00, ?it/s]
2026-05-19 07:14:55.963771: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1779174896.351753     794 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1779174896.484359     794 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1779174897.441436     794 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00

In [10]:
# Inference on the GSM8K test set. Resumable.
!python -m src.inference.generate --condition student_set_c

[auto] using checkpoint: /kaggle/working/COT_lab/outputs/checkpoints/student_set_c/checkpoint-376
[student_set_c] loading model from /kaggle/working/COT_lab/outputs/checkpoints/student_set_c/checkpoint-376
Loading weights: 100%|█| 282/282 [00:00<00:00, 1383.31it/s, Materializing param=
The tied weights mapping and config for this model specifies to tie shared.weight to lm_head.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning
[student_set_c] device=cuda, resuming from record 8
[student_set_c] gen_kwargs={'max_new_tokens': 512, 'do_sample': False, 'num_beams': 4, 'length_penalty': 1.0, 'repetition_penalty': 1.15, 'no_repeat_ngram_size': 4}
student_set_c: 100%|██████████████████████████| 164/164 [22:47<00:00,  8.34s/it]
[student_set_c] done: 1311 examples in 1368s (1.04s/ex) -> /kaggle/working/COT_lab/outputs/generations/student_set_c.jsonl


In [13]:
# Prune to the best checkpoint only (keeps disk under control).
!python -m src.utils.cleanup_ckpts --run-name student_set_c --keep best

--- pruning (keep=best) ---
[student_set_c] 5.6G -> 2.8G   kept: checkpoint-376
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop1       20G  2.9G   17G  15% /kaggle/working


In [2]:
!head -5 /kaggle/working/LAB1/outputs/generations/student_set_c.jsonl


{"question": "Janet\u2019s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?", "generated_cot": "Janet's duck lays 16 eggs per day. This means that she makes 36 dollars in total every morning and 12 hours at the farmers\u2019 market each week, for an average wage of $36/day or $24/2 = $44. She also baked muffin to her friends everyday with 4, which is four times as much money than baking them all together so far this year! We can add these two numbers up by multiplying 4: 48 + 3 * (4*4)= 72 Now we just know how many days it will take until they make their final payment before selling anything else? That would give us $62. So what do you get from making $16 on Wednesday nights when there are only 6 people left over after Friday night?! How long does one person spend d

In [3]:
!cat /kaggle/working/LAB1/outputs/runs/03_train_student_set_c.json

{
  "stage": "03_train",
  "run_name": "student_set_c",
  "started_at": "2026-05-19T07:15:15Z",
  "completed_at": "2026-05-19T07:40:21Z",
  "duration_seconds": 1505.61,
  "status": "completed",
  "config_hash": "sha256:4550f1b46f5f59fc",
  "config_snapshot": {
    "model_name": "google/flan-t5-base",
    "train_file": "data/processed/set_c_calculator.jsonl",
    "n_epochs": 8,
    "limit": null,
    "learning_rate": 5e-05,
    "weight_decay": 0.05,
    "warmup_ratio": 0.1,
    "batch_size": 4,
    "gradient_accumulation_steps": 8,
    "max_input_length": 512,
    "max_target_length": 512,
    "early_stopping_patience": 2,
    "seed": 42
  },
  "inputs": [
    "data/processed/set_c_calculator.jsonl"
  ],
  "outputs": [
    "outputs/checkpoints/student_set_c/checkpoint-334",
    "outputs/checkpoints/student_set_c/checkpoint-376"
  ],
  "metrics": {
    "n_train": 3050,
    "n_val": 339,
    "n_epochs_completed": 7.87434554973822,
    "best_epoch": 6.994764397905759,
    "best_eval_loss":

In [5]:
!head -5 /kaggle/working/LAB1/data/processed/set_c_calculator.jsonl

{"sample_index": 0, "question": "Natalia sold clips to 48 of her friends in April, and then she sold half as many clips in May. How many clips did Natalia sell altogether in April and May?", "cot": "In April, Natalia sold 48 clips. \nIn May, she sold half as many clips. Half of 48 is 24, so she sold 24 clips in May. \n\nIn total, she sold 48 + 24 = 72 clips in April and May.", "gold_answer": "72", "teacher_predicted_answer": "72.", "calculator_corrected_cot": false, "n_calc_edits": 0}
{"sample_index": 3, "question": "Julie is reading a 120-page book. Yesterday, she was able to read 12 pages and today, she read twice as many pages as yesterday. If she wants to read half of the remaining pages tomorrow, how many pages should she read?", "cot": "Yesterday, Julie read 12 pages. \nToday, she read twice as many pages as yesterday, so she read 24 pages. \nThat means she has read a total of 36 pages so far. \nThat means she has 120 - 36 = 84 pages left to read. \nHalf of 84 is 42, so tomorrow 

In [6]:
import json
n_total, n_last_edited = 0, 0
with open('/kaggle/working/LAB1/data/processed/set_c_calculator.jsonl') as f:
    for line in f:
        rec = json.loads(line)
        n_total += 1
        if rec.get('calculator_corrected_cot'):
            n_last_edited += 1
print(f"{n_last_edited}/{n_total} chains were patched")


50/3389 chains were patched


### 4.student_set_a — `student_set_a`

All teacher CoTs (~7,473). Includes the ~55% with wrong final answers — the noisy baseline that motivates the Set B/C filters.

In [ ]:
# Train. Resumable: re-run with --resume after a timeout.
!python -m src.train.finetune \
  --config config/config.yaml \
  --train data/processed/set_a_nofilter.jsonl \
  --run-name student_set_a
  # add --resume after a timeout

In [ ]:
# Inference on the GSM8K test set. Resumable.
!python -m src.inference.generate --condition student_set_a

In [ ]:
# Prune to the best checkpoint only (keeps disk under control).
!python -m src.utils.cleanup_ckpts --run-name student_set_a --keep best

### 4.baseline — `baseline` (no fine-tuning)

Zero-shot inference on `google/flan-t5-base` straight from the Hub. No training step; no checkpoint to prune.

In [ ]:
!python -m src.inference.generate --condition baseline

### 4.eval — Refresh Stage 5a accuracy (base conditions)

Rebuilds `outputs/eval_results/accuracy.csv` from whatever generations currently exist on disk. Safe to run at any time — partial sweeps still produce a partial table.

In [7]:
!python -m src.eval.accuracy \
  --conditions baseline student_direct_ft student_set_a student_set_b student_set_c

/kaggle/working/COT_lab/src/eval/accuracy.py:82: UserWarning: set_ticklabels() should only be used with a fixed number of ticks, i.e. after set_ticks() or using a FixedLocator.
  ax.set_xticklabels(conds, rotation=25, ha="right")
condition                     n      accuracy
-----------------  ------------  ------------
baseline                   1319         4.32%
student_direct_ft          1319         5.23%
student_set_a              1319         2.65%
student_set_b              1319         3.34%
student_set_c              1319         1.82%


### 4.student_set_c_mix — `student_set_c_mix`

Mixed fine-tune: Set C (calculator-patched CoTs) + direct-FT targets. Tests whether blending chain-of-thought with answer-only supervision improves robustness on T5-base.

In [21]:
import json, random
from pathlib import Path

DATA = Path("/kaggle/working/LAB1/data/processed")
random.seed(42)  # deterministic, matches project seed

set_c = (DATA / "set_c_calculator.jsonl").read_text().splitlines()
direct = (DATA / "direct_ft.jsonl").read_text().splitlines()

# Sample direct_ft down to match Set C size
direct_sampled = random.sample(direct, len(set_c))

mixed = set_c + direct_sampled
random.shuffle(mixed)  # interleave so a batch sees both types

out = DATA / "set_c_plus_direct.jsonl"
out.write_text("\n".join(mixed) + "\n")
print(f"Set C: {len(set_c)}")
print(f"direct_ft (sampled): {len(direct_sampled)}")
print(f"Total: {len(mixed)} -> {out}")


Set C: 3389
direct_ft (sampled): 3389
Total: 6778 -> /kaggle/working/COT_lab/data/processed/set_c_plus_direct.jsonl


In [19]:
!wc -l /kaggle/working/LAB1/data/processed/set_c_plus_direct.jsonl

6778 /kaggle/working/COT_lab/data/processed/set_c_plus_direct.jsonl


In [23]:
!head -2 /kaggle/working/LAB1/data/processed/set_c_plus_direct.jsonl 
!tail -2 /kaggle/working/LAB1/data/processed/set_c_plus_direct.jsonl 

{"sample_index": 6118, "question": "A single train car can carry 60 passengers. A 747 airplane can carry 366 passengers. How many more passengers can a train with 16 cars carry than 2 airplanes?", "cot": "A single train car can carry 60 passengers. \nA 747 airplane can carry 366 passengers. \n\nHow many more passengers can a train with 16 cars carry than 2 airplanes?\n\nA train with 16 cars can carry 960 passengers. \n2 airplanes can carry 732 passengers. \n\nA train with 16 cars can carry more than 2 airplanes by 228 passengers.", "gold_answer": "228", "teacher_predicted_answer": "228.", "calculator_corrected_cot": false, "n_calc_edits": 0}
{"sample_index": 1655, "question": "Abe owns a restaurant. Every month he spends a third of his budget on food, a quarter of his budget on restaurant supplies, and the rest of his budget on employee wages. If his budget is $3000 and he uses it up every month, how much is he spending on wages?", "cot": "Abe spends a third of his budget on food. This

In [29]:
!cd /kaggle/working/LAB1 && python -u -m src.train.finetune \
    --train data/processed/set_c_plus_direct.jsonl \
    --run-name student_set_c_mix 2>&1 | tee /kaggle/working/training_log.txt

2026-05-19 10:13:45.090950: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1779185625.113173    1419 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1779185625.119884    1419 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1779185625.137953    1419 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1779185625.138005    1419 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1779185625.138011    1419 computation_placer.cc:177] computation placer alr

In [30]:
!cd /kaggle/working/LAB1 && python -u -m src.inference.generate \
    --condition student_set_c_mix 2>&1 | tee /kaggle/working/inference_log.txt

[auto] using checkpoint: /kaggle/working/COT_lab/outputs/checkpoints/student_set_c_mix/checkpoint-760
[student_set_c_mix] loading model from /kaggle/working/COT_lab/outputs/checkpoints/student_set_c_mix/checkpoint-760
[student_set_c_mix] device=cuda, resuming from record 0
[student_set_c_mix] gen_kwargs={'max_new_tokens': 512, 'do_sample': False, 'num_beams': 4, 'length_penalty': 1.0, 'repetition_penalty': 1.15, 'no_repeat_ngram_size': 4}
student_set_c_mix: 100%|██████████| 165/165 [10:11<00:00,  3.70s/it]
[student_set_c_mix] done: 1319 examples in 611s (0.46s/ex) -> /kaggle/working/COT_lab/outputs/generations/student_set_c_mix.jsonl


In [31]:
!cd /kaggle/working/LAB1 && python -u -m src.eval.accuracy --conditions \
    baseline student_direct_ft student_set_a student_set_b student_set_c student_set_c_mix \
    2>&1 | tee /kaggle/working/accuracy_log.txt

/kaggle/working/COT_lab/src/eval/accuracy.py:82: UserWarning: set_ticklabels() should only be used with a fixed number of ticks, i.e. after set_ticks() or using a FixedLocator.
  ax.set_xticklabels(conds, rotation=25, ha="right")
condition                     n      accuracy
-----------------  ------------  ------------
baseline                   1319         4.32%
student_direct_ft          1319         5.23%
student_set_a              1319         2.65%
student_set_b              1319         3.34%
student_set_c              1319         1.82%
student_set_c_mix          1319         2.81%


In [32]:
!cd /kaggle/working/LAB1 && python -u -m src.eval.receval.score_chain \
    --condition student_set_c_mix 2>&1 | tee /kaggle/working/receval_log.txt

  info scorer: distilgpt2
student_set_c_mix: 100%|██████████| 1319/1319 [03:09<00:00,  6.96it/s]

  1319 examples  190s total  14.4s/100ex
  intra=0.9119  inter=0.1155  info=-2.0352


In [35]:
import json
from pathlib import Path

RUNS = Path("/kaggle/working/LAB1/outputs/runs")
CONDITIONS = [
    # base (FLAN-T5-base)
    "baseline", "student_direct_ft",
    "student_set_a", "student_set_b", "student_set_c",
    "student_set_c_mix",
    # large (FLAN-T5-large)
    "student_direct_ft_large", "student_set_a_large", "student_set_b_large", "student_set_c_large",
]

print(f"{'condition':<28} {'n':>5}  {'intra':>16}  {'inter':>16}  {'info':>18}")
print("-" * 92)
for cond in CONDITIONS:
    card = RUNS / f"05b_{cond}.json"
    if not card.exists():
        print(f"{cond:<28}  (not run)")
        continue
    m = json.loads(card.read_text())["metrics"]
    intra = f"{m['intra_mean']:.3f} +/- {m['intra_std']:.3f}"
    inter = f"{m['inter_mean']:.3f} +/- {m['inter_std']:.3f}"
    info  = f"{m['info_mean']:.3f} +/- {m['info_std']:.3f}"
    print(f"{cond:<28} {m['n_scored']:>5}  {intra:>16}  {inter:>16}  {info:>18}")


condition                  n             intra             inter                info
--------------------------------------------------------------------------------------
baseline                1319     0.968 ± 0.031     0.058 ± 0.198      -1.322 ± 1.076
student_direct_ft       1319     0.912 ± 0.021     0.722 ± 0.361       1.479 ± 1.355
student_set_a           1319     0.921 ± 0.030     0.068 ± 0.191      -2.515 ± 1.708
student_set_b           1319     0.912 ± 0.034     0.061 ± 0.183      -2.421 ± 1.695
student_set_c           1319     0.956 ± 0.033     0.115 ± 0.266      -2.420 ± 1.493
student_set_c_mix       1319     0.912 ± 0.034     0.116 ± 0.268      -2.035 ± 1.951


## 5. Large model crossover — FLAN-T5-large (optional)

Tests **Ho et al.'s central claim** (CoT vs direct FT crossover with scale).

**Scope**: four large runs — `direct_ft`, `set_b`, `set_c`, `set_a`. Includes `set_a_large` to measure whether unfiltered CoT data hurts less when the student model is larger.

**T4 memory math** (FLAN-T5-large is 770M params, ~3 GB fp16):
- batch_size dropped 4 → 2 (fits with optimizer states)
- grad-accum raised 8 → 16 (keeps effective batch at 32)
- learning rate dropped 5e-5 → 3e-5 (standard for larger T5)
- epochs dropped 8 → 4 (large models converge faster; less overfit risk)

**Time budget**: ~3–4 h training + ~25 min inference per condition. Plan for 3 Kaggle sessions. All training is resumable via `--resume`.

**Disk math**: each large checkpoint is ~3 GB. `save_total_limit=2` means ~6 GB live during training. The cleanup cell after each run prunes to ~3 GB (best only).

### 5.1 student_direct_ft_large — `student_direct_ft_large`

Standard fine-tune at T5-large scale.


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.train.finetune \
    --train data/processed/direct_ft.jsonl \
    --run-name student_direct_ft_large \
    --model google/flan-t5-large \
    --batch-size 2 \
    --grad-accum 16 \
    --lr 3e-5 \
    --epochs 4 \
    2>&1 | tee /kaggle/working/train_direct_ft_large.log
# add --resume after a timeout



2026-05-19 08:31:33.720409: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1779179493.743424     800 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1779179493.750666     800 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1779179493.769039     800 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1779179493.769089     800 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1779179493.769096     800 computation_placer.cc:177] computation placer alr

In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.inference.generate \
    --condition student_direct_ft_large \
    2>&1 | tee /kaggle/working/inference_direct_ft_large.log


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.utils.cleanup_ckpts \
    --run-name student_direct_ft_large --keep best \
    2>&1 | tee /kaggle/working/cleanup_direct_ft_large.log


### 5.2 student_set_b_large — `student_set_b_large`

CoT fine-tune at T5-large scale. Compare directly against `student_direct_ft_large` — if CoT wins, you've replicated Ho's crossover.


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.train.finetune \
    --train data/processed/set_b_magister.jsonl \
    --run-name student_set_b_large \
    --model google/flan-t5-large \
    --batch-size 2 \
    --grad-accum 16 \
    --lr 3e-5 \
    --epochs 4 \
    2>&1 | tee /kaggle/working/train_set_b_large.log
# add --resume after a timeout


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.inference.generate \
    --condition student_set_b_large \
    2>&1 | tee /kaggle/working/inference_set_b_large.log


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.utils.cleanup_ckpts \
    --run-name student_set_b_large --keep best \
    2>&1 | tee /kaggle/working/cleanup_set_b_large.log


### 5.3 student_set_c_large — `student_set_c_large`

Calculator-patched CoT fine-tune at T5-large scale. Paired with `student_set_b_large` to test whether arithmetic correction helps when the student is large enough to execute the corrected steps.


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.train.finetune \
    --train data/processed/set_c_calculator.jsonl \
    --run-name student_set_c_large \
    --model google/flan-t5-large \
    --batch-size 2 \
    --grad-accum 16 \
    --lr 3e-5 \
    --epochs 4 \
    2>&1 | tee /kaggle/working/train_set_c_large.log
# add --resume after a timeout


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.inference.generate \
    --condition student_set_c_large \
    2>&1 | tee /kaggle/working/inference_set_c_large.log


In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.utils.cleanup_ckpts \
    --run-name student_set_c_large --keep best \
    2>&1 | tee /kaggle/working/cleanup_set_c_large.log


### 5.4 student_set_a_large — `student_set_a_large`

Unfiltered CoT fine-tune at T5-large scale. Tests whether the no-filter penalty (Set A < baseline at base scale) persists at large scale.

In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.train.finetune \
    --train data/processed/set_a_nofilter.jsonl \
    --run-name student_set_a_large \
    --model google/flan-t5-large \
    --batch-size 2 \
    --grad-accum 16 \
    --lr 3e-5 \
    --epochs 4 \
    2>&1 | tee /kaggle/working/train_set_a_large.log
# add --resume after a timeout

In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.inference.generate \
    --condition student_set_a_large \
    2>&1 | tee /kaggle/working/inference_set_a_large.log

In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.utils.cleanup_ckpts \
    --run-name student_set_a_large --keep best \
    2>&1 | tee /kaggle/working/cleanup_set_a_large.log

### 5.eval — Refresh Stage 5a accuracy (all conditions, base + large)

Rebuilds the accuracy table including the large-model conditions. Compare base vs large head-to-head for direct_ft and set_b.

In [ ]:
!cd /kaggle/working/LAB1 && python -u -m src.eval.accuracy --conditions \
    baseline student_direct_ft student_set_a student_set_b student_set_c student_set_c_mix \
    student_direct_ft_large student_set_a_large student_set_b_large student_set_c_large \
    2>&1 | tee /kaggle/working/accuracy_log.txt


---

## 6. Stage 5b — ReCEval (recommend a fresh session)

Slow: 3-5 hours for the full base sweep. Reads only `outputs/generations/`, so it does not need the checkpoints (you can have pruned them).

**Smoke-test first** on 20 examples to verify the NLI model loads and scoring works end-to-end.

**Info scorer**: `EleutherAI/pythia-1b` by default (used in all experiments). Override with `--info-scorer EleutherAI/pythia-410m` for a lighter scorer.


In [8]:
# Smoke test: 20 examples on student_set_b.
!python -m src.eval.receval.score_chain --condition student_set_b --smoke 20

config.json: 100%|█████████████████████████████| 762/762 [00:00<00:00, 2.87MB/s]
tokenizer_config.json: 100%|██████████████████| 26.0/26.0 [00:00<00:00, 107kB/s]
vocab.json: 1.04MB [00:00, 5.58MB/s]
merges.txt: 456kB [00:00, 3.19MB/s]
tokenizer.json: 1.36MB [00:00, 5.63MB/s]
model.safetensors: 100%|██████████████████████| 353M/353M [00:02<00:00, 131MB/s]
Loading weights: 100%|█| 76/76 [00:00<00:00, 1715.64it/s, Materializing param=tr
GPT2LMHeadModel LOAD REPORT from: distilgpt2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
transformer.h.{0, 1, 2, 3, 4, 5}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
generation_config.json: 100%|███████████████████| 124/124 [00:00<00:00, 572kB/s]
  info scorer: distilgpt2
student_set_b:   0%|                                     | 0/20 [00:00<?, ?it/s]
config.json: 1.05kB [0

### 6.1 ReCEval — base conditions

Scores all six base conditions. Each takes ~3 min on T4.


In [10]:
!python -m src.eval.receval.score_chain --condition baseline
# add --max-examples 500 if a single condition is taking >2h

Loading weights: 100%|█| 76/76 [00:00<00:00, 1607.26it/s, Materializing param=tr
GPT2LMHeadModel LOAD REPORT from: distilgpt2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
transformer.h.{0, 1, 2, 3, 4, 5}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
  info scorer: distilgpt2
Loading weights:   0%|                                  | 0/106 [00:00<?, ?it/s]
Loading weights:   1%| | 1/106 [00:00<00:00, 1081.56it/s, Materializing param=cl
Loading weights:   1%| | 1/106 [00:00<00:00, 878.57it/s, Materializing param=cla
Loading weights:   2%| | 2/106 [00:00<00:00, 1264.49it/s, Materializing param=cl
Loading weights:   2%| | 2/106 [00:00<00:00, 1076.15it/s, Materializing param=cl
Loading weights:   3%| | 3/106 [00:00<00:00, 1206.99it/s, Materializing param=de
Loading weights:   3%| | 3/106 [00:00<00:00, 1032.06it/

In [11]:
!python -m src.eval.receval.score_chain --condition student_set_a
# add --max-examples 500 if a single condition is taking >2h

Loading weights: 100%|█| 76/76 [00:00<00:00, 1621.42it/s, Materializing param=tr
GPT2LMHeadModel LOAD REPORT from: distilgpt2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
transformer.h.{0, 1, 2, 3, 4, 5}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
  info scorer: distilgpt2
Loading weights:   0%|                                  | 0/106 [00:00<?, ?it/s]
Loading weights:   1%| | 1/106 [00:00<00:00, 9686.61it/s, Materializing param=cl
Loading weights:   1%| | 1/106 [00:00<00:00, 1763.05it/s, Materializing param=cl
Loading weights:   2%| | 2/106 [00:00<00:00, 1816.50it/s, Materializing param=cl
Loading weights:   2%| | 2/106 [00:00<00:00, 453.05it/s, Materializing param=cla
Loading weights:   3%| | 3/106 [00:00<00:00, 592.75it/s, Materializing param=deb
Loading weights:   3%| | 3/106 [00:00<00:00, 481.55it/s

In [12]:
!python -m src.eval.receval.score_chain --condition student_set_b
# add --max-examples 500 if a single condition is taking >2h

Loading weights: 100%|█| 76/76 [00:00<00:00, 1696.63it/s, Materializing param=tr
GPT2LMHeadModel LOAD REPORT from: distilgpt2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
transformer.h.{0, 1, 2, 3, 4, 5}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
  info scorer: distilgpt2
Loading weights:   0%|                                  | 0/106 [00:00<?, ?it/s]
Loading weights:   1%| | 1/106 [00:00<00:00, 8738.13it/s, Materializing param=cl
Loading weights:   1%| | 1/106 [00:00<00:00, 987.82it/s, Materializing param=cla
Loading weights:   2%| | 2/106 [00:00<00:00, 1056.63it/s, Materializing param=cl
Loading weights:   2%| | 2/106 [00:00<00:00, 564.70it/s, Materializing param=cla
Loading weights:   3%| | 3/106 [00:00<00:00, 730.33it/s, Materializing param=deb
Loading weights:   3%| | 3/106 [00:00<00:00, 588.73it/s

In [9]:
!python -m src.eval.receval.score_chain --condition student_set_c
# add --max-examples 500 if a single condition is taking >2h

Loading weights: 100%|█| 76/76 [00:00<00:00, 1695.99it/s, Materializing param=tr
GPT2LMHeadModel LOAD REPORT from: distilgpt2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
transformer.h.{0, 1, 2, 3, 4, 5}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
  info scorer: distilgpt2
Loading weights:   0%|                                  | 0/106 [00:00<?, ?it/s]
Loading weights:   1%| | 1/106 [00:00<00:00, 9039.45it/s, Materializing param=cl
Loading weights:   1%| | 1/106 [00:00<00:00, 1513.64it/s, Materializing param=cl
Loading weights:   2%| | 2/106 [00:00<00:00, 1024.88it/s, Materializing param=cl
Loading weights:   2%| | 2/106 [00:00<00:00, 893.64it/s, Materializing param=cla
Loading weights:   3%| | 3/106 [00:00<00:00, 964.65it/s, Materializing param=deb
Loading weights:   3%| | 3/106 [00:00<00:00, 876.86it/s

### 6.2 ReCEval — large conditions (only if you ran section 5)

Scores `direct_ft_large`, `set_a_large`, `set_b_large`, `set_c_large`. Each takes ~3 min on T4.


In [ ]:
!python -u -m src.eval.receval.score_chain --condition student_direct_ft_large 2>&1 | tee /kaggle/working/receval_direct_ft_large.log
# add --max-examples 500 if a single condition is taking >2h


In [ ]:
!python -u -m src.eval.receval.score_chain --condition student_set_a_large 2>&1 | tee /kaggle/working/receval_set_a_large.log
# add --max-examples 500 if a single condition is taking >2h


In [ ]:
!python -m src.eval.receval.score_chain --condition student_set_b_large
# add --max-examples 500 if a single condition is taking >2h

In [ ]:
!python -u -m src.eval.receval.score_chain --condition student_set_c_large 2>&1 | tee /kaggle/working/receval_set_c_large.log
# add --max-examples 500 if a single condition is taking >2h



In [ ]:
import json
from pathlib import Path

RUNS = Path("/kaggle/working/LAB1/outputs/runs")
CONDITIONS = [
    # base (FLAN-T5-base, 250M)
    "baseline",
    "student_direct_ft",
    "student_set_a",
    "student_set_b",
    "student_set_c",
    "student_set_c_mix",
    # large (FLAN-T5-large, 770M)
    "student_direct_ft_large",
    "student_set_a_large",
    "student_set_b_large",
    "student_set_c_large",
]

print(f"{'condition':<28} {'n':>5}  {'intra':>16}  {'inter':>16}  {'info':>18}")
print("-" * 92)
for cond in CONDITIONS:
    card = RUNS / f"05b_{cond}.json"
    if not card.exists():
        print(f"{cond:<28}  (not run)")
        continue
    m = json.loads(card.read_text())["metrics"]
    intra = f"{m['intra_mean']:.3f} ± {m['intra_std']:.3f}"
    inter = f"{m['inter_mean']:.3f} ± {m['inter_std']:.3f}"
    info  = f"{m['info_mean']:.3f} ± {m['info_std']:.3f}"
    print(f"{cond:<28} {m['n_scored']:>5}  {intra:>16}  {inter:>16}  {info:>18}")


## 7. Project status

In [ ]:
!python -m src.status

## 8. Push results to `main`

Force-adds the small output artifacts (run-cards, eval CSVs, per-example JSONLs, plots, generations) — **not** model checkpoints. Pulls `main` with `-X ours` so on any conflict, Kaggle's version wins.

Safe to re-run between blocks if you want to checkpoint progress.


In [42]:
import datetime, subprocess

ARTIFACTS = [
    'outputs/runs',
    'outputs/eval_results',
    'outputs/plots',
    'outputs/generations',
]

def _run(*args):
    return subprocess.run(['git', '-C', REPO_DIR, *args],
                          capture_output=True, text=True, check=False)

# 1. Force-add (outputs/ is gitignored)
for path in ARTIFACTS:
    if os.path.isdir(os.path.join(REPO_DIR, path)):
        _run('add', '-f', path)

# 2. Bail out cleanly if nothing changed
status = _run('status', '--short').stdout
if not status.strip():
    print('Nothing to commit.')
else:
    msg = f'kaggle: results @ {datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M UTC")}'
    print('staged:\n' + status)
    r = _run('commit', '-m', msg)
    print(r.stdout or r.stderr)

    # 3. Pull with merge strategy that prefers Kaggle on conflict
    _run('fetch', 'origin')
    r = _run('pull', '--no-edit', '-X', 'ours', 'origin', BRANCH)
    print(r.stdout or r.stderr)

    # 4. Push
    r = _run('push', 'origin', BRANCH)
    print(r.stdout or r.stderr)
    print('done.')

/tmp/ipykernel_57/420135657.py:24: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  msg = f'kaggle: results @ {datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M UTC")}'


staged:
 M scripts/01_download.sh
 M scripts/02_filter.sh
 M scripts/03_train_direct_ft.sh
 M scripts/03_train_set_a.sh
 M scripts/03_train_set_b.sh
 M scripts/03_train_set_c.sh
 M scripts/04_inference.sh
 M scripts/05a_accuracy.sh
 M scripts/05b_receval.sh
 M scripts/status.sh

On branch dev
Your branch is up to date with 'origin/dev'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   scripts/01_download.sh
	modified:   scripts/02_filter.sh
	modified:   scripts/03_train_direct_ft.sh
	modified:   scripts/03_train_set_a.sh
	modified:   scripts/03_train_set_b.sh
	modified:   scripts/03_train_set_c.sh
	modified:   scripts/04_inference.sh
	modified:   scripts/05a_accuracy.sh
	modified:   scripts/05b_receval.sh
	modified:   scripts/status.sh

no changes added to commit (use "git add" and/or "git commit -a")

Already up to date.

Everything up-to-date

done.


## 9. Aggressive cleanup (end of session)

Run **after** the push cell — once results are on `dev`, the local checkpoints and HF cache are disposable.

- `cleanup_ckpts --all --keep none` deletes every fine-tune checkpoint (~6-15 GB freed depending on what you trained).
- The next cell wipes the HF download cache and pip wheels (~2 GB).

**Don't run this if you still plan to run Stage 4 inference** for a checkpoint — pruning to `none` makes it unrecoverable without retraining.

In [ ]:
!python -m src.utils.cleanup_ckpts --all --keep none

In [ ]:
import os, shutil, subprocess

TARGETS = [
    os.path.expanduser('~/.cache/huggingface'),
    os.path.expanduser('~/.cache/pip'),
]

def _du(path):
    if not os.path.exists(path):
        return '   -   '
    r = subprocess.run(['du', '-sh', path], capture_output=True, text=True)
    return r.stdout.split('\t')[0].strip()

print('--- before ---')
for p in TARGETS:
    print(f'  {_du(p):>8}  {p}')
subprocess.run(['df', '-h', '/kaggle/working'])

for p in TARGETS:
    shutil.rmtree(p, ignore_errors=True)

print('\n--- after ---')
for p in TARGETS:
    print(f'  {_du(p):>8}  {p}')
subprocess.run(['df', '-h', '/kaggle/working'])
print('\ndone.')